File: src/pi/modules/control_tasks/valve_control.py
import time
from modules.mcl.flag import Flag
from modules.lib.helpers import enqueue
from modules.mcl.registry import Registry
from modules.lib.packet import Log, LogPriority
from modules.lib.enums import ActuationType, ValvePriority, ValveType, ValveLocation

class ValveControl():
    def __init__(self, registry: Registry, flag: Flag):
        # Initialize the ValveControl class
        self.registry = registry
        self.flag = flag

    
    def begin(self, config: dict):
        # Create basic varibales
        self.config = config
        self.valves = self.config["valves"]["list"]
        self.send_interval = self.config["valves"]["send_interval"]
        self.last_send_time = None
        # NOTE: Ground states are the same as abort states, meaning a valve's "natural" state (which is in the config) is the same as it's abort state
        # This is so that if the power dies for some reason, the valves will automatically abort
        self.aborted = False
        natural_to_actuation = {"OPEN": ActuationType.OPEN_VENT, "CLOSED": ActuationType.CLOSE_VENT}
        self.abort_actuations = {}
        for valve_loc in self.valves["solenoid"]:
            natural_str = self.valves["solenoid"][valve_loc]["natural"]
            natural = natural_to_actuation[natural_str]
            self.abort_actuations[valve_loc] = natural


    def send_valve_data(self):
        # Send the valve data to GS, basically a dict with the format: {solenoid: {loc1: "open", loc2: "closed"}}
        message = {}
        for valve_type in self.valves:
            message[valve_type] = {}
            for valve_loc in self.valves[valve_type]:
                _, val, _ = self.registry.get(("valve", valve_type, valve_loc))
                message[valve_type][valve_loc] = val
        log = Log(header="valve_data", message=message)
        enqueue(self.flag, log, LogPriority.INFO)


    def abort(self):
        # Loop through all the solenoid valves
        for valve_loc in self.valves["solenoid"]:
            actuation_type = self.registry.get(("valve_actuation", "actuation_type", ValveType.SOLENOID, valve_loc))[1]
            actuation_priority = self.registry.get(("valve_actuation", "actuation_priority", ValveType.SOLENOID, valve_loc))[1]
            # If the valve hasn't been aborted, then abort it.
            if actuation_priority != ValvePriority.ABORT_PRIORITY:
                self.flag.put(("solenoid", "actuation_type", valve_loc), self.abort_actuations[valve_loc])
                self.flag.put(("solenoid", "actuation_priority", valve_loc), ValvePriority.ABORT_PRIORITY)


    def undo_abort(self):
        # Loop through all the solenoid valves
        for valve_loc in self.valves["solenoid"]:
            actuation_priority = self.registry.get(("valve_actuation", "actuation_priority", ValveType.SOLENOID, valve_loc))[1]
            # This if statement should technically be true for all valves, but basically it just makes sure that the valve aborted.
            if actuation_priority == ValvePriority.ABORT_PRIORITY:
                # Reset the valve priority to zero. This means that whatever the Pi tells it to do next, it'll do that, instead of sticking to its aborted state.
                self.flag.put(("solenoid", "actuation_type", valve_loc), ActuationType.NONE)
                self.flag.put(("solenoid", "actuation_priority", valve_loc), ValvePriority.NONE)


    def check_abort(self):
        # If you just got a message to abort, then abort
        if self.registry.get(("general", "soft_abort"))[1] and not self.aborted:
            self.abort()
            self.aborted = True
        # If you just got a message to undo abort, then undo abort
        elif not self.registry.get(("general", "soft_abort"))[1] and self.aborted:
            self.undo_abort()
            self.aborted = False


    def execute(self):
        # Check if the rocket has been aborted (and act correspondingly)
        self.check_abort()
        # Send valve data every once in a while to update GS
        if self.last_send_time is None or time.time() - self.last_send_time > self.send_interval:
            self.send_valve_data()
            self.last_send_time = time.time()
        

File: src/modules/control_tasks/ValveControl.cpp
//
// Created by adiv413 on 4/24/2020.
//

#include <flight/modules/control_tasks/ValveControl.hpp>
#include <flight/modules/lib/Enums.hpp>
#include <flight/modules/mcl/Config.hpp>
#include <flight/modules/mcl/Registry.hpp>
#include <flight/modules/mcl/Flag.hpp>
#include <Logger/logger_util.h>
#include <chrono>
#include <string>

ValveControl::ValveControl() {
    this->send_interval = global_config.valves.send_interval;
    this->last_send_time = 0;
}

void ValveControl::begin() {
    log("Valve Control: Beginning");
    global_flag.log_info("response",R"({"header": "info", "Description": "Valve Control started"})");
}

void ValveControl::execute() {
    check_abort();
    auto current_time = chrono::system_clock::now().time_since_epoch().count();

    if (last_send_time == 0 || current_time > last_send_time + send_interval) {
        send_valve_data();
        last_send_time = current_time;
    }
}

void ValveControl::send_valve_data() {
    json valve_data_json = json::object();

    for (const auto& type_pair : global_registry.valves) {
        string type = type_pair.first;
        for (const auto& location_pair : type_pair.second) {
            string location = location_pair.first;
            RegistryValveInfo valve_info = location_pair.second;
            valve_data_json[type + "." + location] = json{
                {"state", solenoid_state_map.at(valve_info.state)},
                {"actuation_type", actuation_type_inverse_map.at(valve_info.actuation_type)}
            };
        }
    }

    global_flag.log_info("valve_data", valve_data_json);
    last_send_time = chrono::system_clock::now().time_since_epoch().count();
}

void ValveControl::abort() {
    for (const auto& valve_ : valves) {
        RegistryValveInfo valve_info = global_registry.valves[valve_.first][valve_.second];

        auto actuation_type = valve_info.actuation_type;
        auto actuation_priority = valve_info.actuation_priority;

        if (actuation_type != ActuationType::OPEN_VENT || actuation_priority != ValvePriority::ABORT_PRIORITY) {
            FlagValveInfo &valve_flag = global_flag.valves[valve_.first][valve_.second];
            valve_flag.actuation_type = ActuationType::OPEN_VENT;
            valve_flag.actuation_priority = ValvePriority::ABORT_PRIORITY;
        }
    }
}

// Set the actuation type to NONE, with ABORT_PRIORITY priority
void ValveControl::undo_abort() {
    for (const auto& valve_ : valves) {
        RegistryValveInfo valve_info = global_registry.valves[valve_.first][valve_.second];

        if (valve_info.actuation_priority == ValvePriority::ABORT_PRIORITY) {
            FlagValveInfo &valve_flag = global_flag.valves[valve_.first][valve_.second];
            valve_flag.actuation_type = ActuationType::NONE;
            valve_flag.actuation_priority = ValvePriority::ABORT_PRIORITY;
        }
    }
}

void ValveControl::check_abort() {
    if (global_registry.general.hard_abort || global_registry.general.soft_abort) {
        abort();
    } else if (!global_registry.general.soft_abort) {
        undo_abort();
    }
}